### DISCUSSION TITLE
[BUG] Trackpad gestures stop working after some time

### DISCUSSION THREAD

**Discussion by arham-saeed** [rxhanson/Charmstone-Community#34] [Q&A]
Using the app for a few months now, everything works great except that trackpad gestures stop working after some time. I can't reproduce the conditions which cause this bug to happen but it happens at least 1-2 times every day. 

The app itself is responsive and the keyboard gestures are also working so the issue seems to be only in the trackpad gestures.

Usually I just kill the app and restart it to solve this but it is becoming quite annoying. I would greatly appreciate any help I can get to solve this. 

charmstone: 0.29
mac os: 13.4.1
Macbook pro 16-inch, 2021 (Apple M1 Pro)

**Comment by rxhanson** [comment]
Thanks for reporting. If you can figure out anything that could be causing this, that would be our best bet on getting some sort of fix in. The gesture API in macOS is a stream of events, and that stream can be affected by many different factors. Typically, the stream will send an error event when the stream is no longer going to publish events, but in your case that's not happening and Charmstone has no way to know that it should try to restart listening to events. The most common cause of event interruption like this is waking from sleep, but it can be caused by other apps that listen to gesture events (like BetterTouchTool). 

As an aside, I was hesitant to add the trackpad gestures into Charmstone for this reason, but the issue appears to affect an incredibly small percentage of users (sorry you're one of them). This leads me to believe that there is something peculiar about your setup that is affecting the event stream in some way. If you can narrow anything down, please let me know.

**Comment by F1248** [comment]
I have exactly the same problem: a few times a day, Charmstone no longer responds to trackpad gestures. After restarting the app, everything works again. I can't reproduce the problem either.
macOS 13.4.1 (c) (22F770820d)
MacBook Air M1 (MacBookAir10,1)
Charmstone 0.29

**Comment by grassybeetle144** [comment]
I also experience this problem twice/three times a day. I often find that the trackpad gesture breaks after waking from sleep, but I don't know how to reproduce it. I find this gesture super useful and would also greatly appreciate a solution :)

macOS 13.4.1
MacBook Pro 14-inch, 2021 (Apple M1 Pro)
Charmstone 0.29

**Comment by rxhanson** [comment]
I'll put in an automatic relaunch on wake in the next release.

**Comment by rxhanson** [comment]
The latest release, v0.30, now has an option for automatically relaunching the app on wake. I was going to make this enabled by default, but making an app relaunch itself is way more complicated and error prone than one might think. In a future update, I will likely make this enabled by default once I can confirm that it's working well.

To enable it, click on the info button next to the Trackpad Gesture section header and check the box for relaunch on wake.

Please try it out and let me know how it goes.

**Reply by arham-saeed** [reply]
haven't had the problem again after enabling the option. Thanks for the fix 🙏 .

**Reply by F1248** [reply]
I didn't have the problem anymore either, thanks for this update as well!

**Reply by rxhanson** [reply]
Great, thanks! I'll plan on making this enabled by default when a gesture is configured, likely in the next release.
